src/plonk/prover.rs

```diff
@@ -98,13 +120,127 @@ impl<C: CurveAffine> Proof<C> {
             })
             .collect();
 
+        // Sample x_0 challenge
+        let x_0: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        // Sample x_1 challenge
+        let x_1: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        // Compute permutation product polynomial commitment
+        let mut permutation_product_polys = vec![];
+        let mut permutation_product_cosets = vec![];
+        let mut permutation_product_cosets_inv = vec![];
+        let mut permutation_product_commitments = vec![];
+        let mut permutation_product_blinds = vec![];
+
+        // Iterate over each permutation
+        for (wires, permuted_values) in srs.meta.permutations.iter().zip(srs.permutations.iter()) {
+            // Goal is to compute the fraction
+            //
+            // (p_j(\omega^i) + \delta^j \omega^i \beta + \gamma) /
+            // (p_j(\omega^i) + \beta s_j(\omega^i) + \gamma)
+            //
+            // where p_j(X) is the jth advice wire in this permutation,
+            // and i is the ith row of the wire.
+            let mut modified_advice = Vec::with_capacity(wires.len());
+
+            // Iterate over each wire of the permutation
+            for (wire, permuted_wire_values) in wires.iter().zip(permuted_values.iter()) {
+                // Grab the advice wire's values from the witness
+                let mut tmp_advice_values = witness.advice[wire.0].clone();
+
+                // For each row i, compute
+                // p_j(\omega^i) + \beta s_j(\omega^i) + \gamma
+                // where p_j(omega^i) = tmp[i]
+                for (tmp_advice_value, permuted_advice_value) in tmp_advice_values
+                    .iter_mut()
+                    .zip(permuted_wire_values.iter())
+                {
+                    *tmp_advice_value += &(x_0 * permuted_advice_value);
+                    *tmp_advice_value += &x_1;
+                }
+                modified_advice.push(tmp_advice_values);
+            }
+
+            // Batch invert to obtain the denominators for the permutation product
+            // polynomial
+            for v in &mut modified_advice {
+                C::Scalar::batch_invert(v);
+            }
```

Comment:
Can we batch invert all of the `v`s together?

Reply:
Yes, there are a few places scattered around the code where such inversions can be folded together. We could try to write a generic Montgomery's trick API so that we can write

```rust
modified_advice.iter_mut().flat_map(|v| v.iter_mut()).batch_invert();
```

or something like that.

Reply:
`modified_advice` is a vector of vectors, so i think we have to break it up into vectors to pass into `batch_invert()`

Reply:
Done!